specs/LaunchingRegisteredProtocols.md

```diff
@@ -0,0 +1,318 @@
+Cancelling the launch gives the host the opportunity to hide the default dialog, display a custom dialog, and then launch the external protocol themselves.
+
+# Description
+
+This event will be raised before the external protocol launch occurs.
+When an attempt to launch an external protocol is made, the default dialog is displayed in which the user can select `Open` or `Cancel` if the host does not cancel the event.  
+The `NavigationStarting` event will be raised before the `LaunchingExternalProtocol` event, followed by the `NavigationCompleted` event.
+The `SourceChanged`, `ContentLoading`, and `HistoryChanged` events will not be raised when a request is made to launch an external protocol.
+
+The `LaunchingExternalProtocol` event will be raised on the `CoreWebView2` interface.
+
+# Examples
+
+## Win32 C++
+    
+```cpp 
+
+AppWindow* m_appWindow;
+wil::com_ptr<ICoreWebView2> m_webView;
+EventRegistrationToken m_launchingExternalProtocolToken = {};
+
+void RegisterLaunchingExternalProtocolHandler()
+{
+    auto webView16 = m_webView.try_query<ICoreWebView2_16>();
+    if (webView16)
+    {
+        CHECK_FAILURE(webView16->add_LaunchingExternalProtocol(
+            Callback<ICoreWebView2LaunchingExternalProtocolEventHandler>(
+                [this](
+                    ICoreWebView2* sender,
+                    ICoreWebView2LaunchingExternalProtocolEventArgs* args)
+                {
+                    auto showDialog = [this, args]
+                    {
+                        // Set the `Cancel` property to `TRUE`, as we will either silently launch the 
+                        // trusted app or display a custom dialog. 
+                        args->put_Cancel(true);
+                        wil::unique_cotaskmem_string uri;
+                        CHECK_FAILURE(args->get_Uri(&uri));
+                        if (wcsicmp(uri.get(), L"calculator://") == 0)
+                        {
+                            // If this matches our desired protocol, launch the
+                            // calculator app.
+                            std::wstring protocol_url = L"calculator://";
+                            SHELLEXECUTEINFO info = {sizeof(info)};
+                            info.fMask = SEE_MASK_NOASYNC;
+                            info.lpVerb = L"open";
+                            info.lpFile = protocol_url.c_str();
+                            info.nShow = SW_SHOWNORMAL;
+                            ::ShellExecuteEx(&info);
+                        }
+                        else
+                        {
+                            // To display a custom dialog we cancel the launch, display
+                            // a custom dialog, and then manually launch the external protocol. 
+                            wil::unique_cotaskmem_string initiating_origin;
+                            CHECK_FAILURE(args->get_InitiatingOrigin(&initiating_origin));

```

Comment:
If the protocol launch comes from a frame, does the event get raised on the root WebView2? I assume so, and that's why we give you the InitiatingOrigin - that is the Source of the frame that is trying to launch the external protocol.

Is there a possibility that we may add this event to the frame objects later?

Reply:
If the protocol launch is initiated from a frame, the event is raised on the root WebView2. It is a possibility that we add this event to the frame object later but no work is planned as of now.

Reply:
Overall this is about applying consistent security and UI to launching other apps. Requiring the dev to subscribe to the event on every frame that gets created would be cumbersome for this purpose. If we want to let devs make per-frame decisions I'd rather add the CoreWebView2Frame object to the existing event args than add the event to CoreWebView2Frame and have to worry about the ordering or bubbling of the events.

Reply:
No change required here.